File: app/src/main/java/com/example/bookcleartool/MainActivity.java
package com.example.bookcleartool;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.style.ForegroundColorSpan;
import android.text.style.RelativeSizeSpan;
import android.util.Pair;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.PermissionChecker;

import com.example.bookcleartool.databinding.ActivityMainBinding;
import com.example.bookcleartool.setting.SettingActivity;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MainActivity extends AppCompatActivity {
    private static final int REQUEST_CODE_STORAGE = 1;
    private static final String[] PERMISSIONS_STORAGE = {Manifest.permission.READ_EXTERNAL_STORAGE};
    private List<CharSequence> list = new ArrayList();
    private float[] hsv = {1, 1, 1};
    private ProgressDialog progressDialog;
    private ClearTask task;
    private ActivityMainBinding mainViewBinding;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mainViewBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mainViewBinding.getRoot());
        if (!verifyStoragePermission()) {
            finish();
            return;
        }
        if (getIntent() != null && getIntent().getData() != null) {
            String path = Utils.turnUri2FilePath(this, getIntent().getData());
            task = new ClearTask();
            task.execute(path);
        } else {
            mainViewBinding.cardViewError.setVisibility(View.VISIBLE);
        }
        mainViewBinding.btnExit.setOnClickListener(v -> {
            finish();
        });

        progressDialog = new ProgressDialog(this);
        progressDialog.setCancelable(false);
        progressDialog.setMessage("处理中……");

        mainViewBinding.listview.setAdapter(new ArrayAdapter<CharSequence>(this, android.R.layout.simple_list_item_1, list));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_setting:
                startActivity(new Intent(this, SettingActivity.class));
                break;
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private boolean verifyStoragePermission() {
        //1.检测权限
        int permission = PermissionChecker.PERMISSION_GRANTED;
        for (String s : PERMISSIONS_STORAGE) {
            permission = ActivityCompat.checkSelfPermission(this, s);
        }
        if (permission != PermissionChecker.PERMISSION_GRANTED) {
            //2.没有权限，弹出对话框申请
            ActivityCompat.requestPermissions(
                    this,
                    PERMISSIONS_STORAGE,
                    REQUEST_CODE_STORAGE
            );
            return false;
        }
        return true;
    }

    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE_STORAGE) {
            int result = PermissionChecker.PERMISSION_GRANTED;
            for (int grantResult : grantResults) {
                result = result ^ grantResult;
            }
            if (result == PermissionChecker.PERMISSION_GRANTED) {
                //权限申请成功
                Toast.makeText(this, "授权成功", Toast.LENGTH_SHORT).show();
            } else {
                //权限申请失败
                Toast.makeText(this, "未获取到所有授权，退出。", Toast.LENGTH_SHORT).show();
                finish();
            }
        }
    }

    private Pair<String, CharSequence> clearText2(String str) {
        if (str == null || str.isEmpty())
            return null;
        Matcher matcher = Pattern.compile("&#\\d*").matcher(str);
        ArrayList<Integer> list1 = new ArrayList<Integer>();
        SpannableString spannableString = new SpannableString(str);
        boolean hasMatched = false;
        while (matcher.find()) {
            hasMatched = true;
            list1.add(matcher.start());
            list1.add(matcher.end());
            spannableString.setSpan(new ForegroundColorSpan(getRandomColor()), matcher.start(), matcher.end(), Spannable.SPAN_INCLUSIVE_EXCLUSIVE);
        }
        if (hasMatched)
            return new Pair(matcher.replaceAll(""), spannableString);
        else
            return null;
    }

    private int getRandomColor() {
        hsv[0] = (float) (Math.random() * 360);
        return Color.HSVToColor(hsv);
    }


    class ProgressBeen {
        int curLine;
        CharSequence str;

        public ProgressBeen(int curLine, CharSequence str) {
            this.curLine = curLine;
            this.str = str;
        }

        public CharSequence toCharSequence() {
            SpannableStringBuilder result = new SpannableStringBuilder();
            SpannableString ss = new SpannableString(String.valueOf(curLine));
            ss.setSpan(new RelativeSizeSpan(.7f), 0, ss.length(), Spannable.SPAN_INCLUSIVE_EXCLUSIVE);
            result.append(ss);
            result.append(str);
            return result;
        }
    }

    class ClearTask extends AsyncTask<String, ProgressBeen, Boolean> {
        public Boolean doInBackground(String... params) {
            final String path = params[0];
            BufferedReader reader = null;
            BufferedWriter writer = null;

            String backPath = path.substring(0, path.lastIndexOf('.')) + '1' + path.substring(path.lastIndexOf('.'));

            try {
                reader = new BufferedReader(new InputStreamReader(new FileInputStream(path), "GBK"));
                writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(backPath), "GBK"));

                String temp;
                Pair<String, CharSequence> pair;
                int lineNum = 0;
                do {
                    temp = reader.readLine();
                    if (temp == null)
                        break;
                    lineNum++;
                    pair = clearText2(temp);
                    if (pair != null) {
                        writer.write(pair.first);
                        publishProgress(new ProgressBeen(lineNum, pair.second));
                    } else
                        writer.write(temp);
                    writer.newLine();
                } while (true);

            } catch (IOException e) {
                publishProgress(new ProgressBeen(-1, e.getMessage()));
                return false;
            } finally {
                if (reader != null) {
                    try {
                        reader.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                if (writer != null) {
                    try {
                        writer.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }

            return true;
        }

        @Override
        public void onPreExecute() {
            super.onPreExecute();
            progressDialog.show();
        }

        @Override
        public void onPostExecute(Boolean result) {
            super.onPostExecute(result);
            progressDialog.dismiss();
            mainViewBinding.listview.setVisibility(result != null && result ? View.VISIBLE : View.GONE);
            mainViewBinding.cardViewError.setVisibility(result != null && result ? View.GONE : View.VISIBLE);

        }

        @Override
        public void onProgressUpdate(ProgressBeen... values) {
            super.onProgressUpdate(values);
            ((ArrayAdapter) mainViewBinding.listview.getAdapter()).notifyDataSetChanged();
            list.add(values[0].toCharSequence());
        }

        @Override
        public void onCancelled() {
            super.onCancelled();
            progressDialog.dismiss();
        }
    }
}

File: app/src/main/java/com/example/bookcleartool/MainActivity.kt
package com.example.bookcleartool

import android.Manifest
import android.app.ProgressDialog
import android.graphics.Color
import android.os.AsyncTask
import android.os.Bundle
import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.style.ForegroundColorSpan
import android.text.style.RelativeSizeSpan
import android.view.View
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.PermissionChecker
import java.io.*
import java.security.Permissions
import java.util.regex.Pattern

class MainActivity : AppCompatActivity() {


    companion object {
        const val REQUEST_CODE_STORAGE = 1
        val PERMISSIONS_STORAGE = arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
    }

    val list = mutableListOf<CharSequence>()

    val progressDialog by lazy {
        ProgressDialog(this).apply { this.setCancelable(false) }.apply { this.setMessage("处理中……") }
    }
    val mListView: ListView by lazy {
        this.findViewById<ListView>(R.id.listview).apply {
            this.adapter =
                ArrayAdapter<CharSequence>(this.context, android.R.layout.simple_list_item_1, list)
        }
    }
    private val tvError: View by lazy { findViewById<View>(R.id.tvError) }
    private val cardViewError: View by lazy { findViewById<View>(R.id.cardViewError) }

    private var task: ClearTask? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (!verifyStoragePermission()) {
            finish()
            return
        }

        if (intent != null && intent.data != null) {
            val path = Utils.turnUri2FilePath(this, intent.data)
            task = ClearTask()
            task!!.execute(path)
        } else {
            cardViewError.visibility = View.VISIBLE
        }
        findViewById<View>(R.id.btn_exit).setOnClickListener { finish() }
    }

    private fun verifyStoragePermission(): Boolean {
        //1.检测权限
        var permission = PermissionChecker.PERMISSION_GRANTED
        for (s in PERMISSIONS_STORAGE) {
            permission = permission or ActivityCompat.checkSelfPermission(this, s)
        }
        if (permission != PermissionChecker.PERMISSION_GRANTED) {
            //2.没有权限，弹出对话框申请
            ActivityCompat.requestPermissions(
                this,
                PERMISSIONS_STORAGE,
                REQUEST_CODE_STORAGE
            )
            return false
        }
        return true
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String?>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_STORAGE) {
            var result = PermissionChecker.PERMISSION_GRANTED
            for (grantResult in grantResults) {
                result = result or grantResult
            }
            if (result == PermissionChecker.PERMISSION_GRANTED) {
                //权限申请成功
                Toast.makeText(this, "授权成功", Toast.LENGTH_SHORT).show()
            } else {
                //权限申请失败
                Toast.makeText(this, "未获取到所有授权，退出。", Toast.LENGTH_SHORT).show()
                finish()
            }
        }
    }
//    private fun clearText(str: String?, ss: SpannableString? = null): Pair<String?, CharSequence?>? {
//        if (str == null || str.isEmpty())
//            return Pair(str, null)
//
//        var start = -1
//        for (i in str.indices) {
//            val c = str[i]
//            if (start == -1) {
//                if (c == '&') start = i
//            } else if (i - start == 1) {
//                if (c != '#') start = -1
//            } else if (i - start > 1) {
//                if (!(c.isDigit() || c === '#' || c === '&')){
//                    var spannableString :SpannableString? = ss
//                    if (spannableString==null)
//                        spannableString= SpannableString(str)
//                    spannableString.setSpan(ForegroundColorSpan(Color.RED),)
//                    return clearText("${str.substring(0, start)}${str.substring(i)}", spannableString)
//                }
//            }
//        }
//
//        return if (start != -1) {
//            Pair(str.substring(0, start), ss)
//        } else
//            Pair(str, ss)
//    }

    private fun clearText2(str: String?): Pair<String, CharSequence>? {
        if (str == null || str.isEmpty())
            return null
        val matcher = Pattern.compile("&#\\d*").matcher(str)
        val list = mutableListOf<Int>()
        val spannableString = SpannableString(str)
        var hasMatched = false
        while (matcher.find()) {
            hasMatched = true
            list.add(matcher.start())
            list.add(matcher.end())
            spannableString.setSpan(
                ForegroundColorSpan(getRandomColor()),
                matcher.start(),
                matcher.end(),
                Spannable.SPAN_INCLUSIVE_EXCLUSIVE
            )
        }

        return if (hasMatched)
            Pair(matcher.replaceAll(""), spannableString)
        else
            null
    }

    private fun getRandomColor(): Int {
        Math.random()
        val hsv = FloatArray(3)
        hsv[0] = Math.random().toFloat() * 360
        hsv[1] = 1f
        hsv[2] = 1f
        return Color.HSVToColor(hsv)
    }

    private inner class ProgressBeen(open val curLine: Int, open val str: CharSequence) {
        override fun toString(): String {
            return "$curLine $str"
        }

        fun toCharSequence(): CharSequence {
            val result = SpannableStringBuilder()
            val ss = SpannableString(curLine.toString())
            ss.setSpan(RelativeSizeSpan(.7f), 0, ss.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
            result.append(ss)
            result.append(str)
            return result
        }
    }

    private inner class ClearTask : AsyncTask<String, ProgressBeen, Boolean>() {
        override fun doInBackground(vararg params: String?): Boolean {
            val path = params[0]
            var reader: BufferedReader? = null
            var writer: BufferedWriter? = null

            val backPath = "${path?.substring(0, path?.lastIndexOf('.'))}1${path?.substring(
                path?.lastIndexOf('.')
            )}"
            try {
                reader = BufferedReader(InputStreamReader(FileInputStream(path), "GBK"))
                writer = BufferedWriter(OutputStreamWriter(FileOutputStream(backPath), "GBK"))

                var temp: String?
                var pair: Pair<String, CharSequence>?
                var lineNum = 0
                do {
                    temp = reader.readLine()
                    if (temp == null)
                        break
                    lineNum++
                    pair = clearText2(temp)
                    if (pair != null) {
                        writer.write(pair.first)
                        publishProgress(ProgressBeen(lineNum, pair.second))
                    } else
                        writer.write(temp)
                    writer.newLine()
                } while (true)

            } catch (e: IOException) {
                publishProgress(ProgressBeen(-1, e.message as CharSequence))
                return false
            } finally {
                reader?.close()
                writer?.close()
            }

            return true
        }

        override fun onPreExecute() {
            super.onPreExecute()
            progressDialog.show()
        }

        override fun onPostExecute(result: Boolean) {
            super.onPostExecute(result)
            progressDialog.dismiss()
            mListView.visibility = if (result) View.VISIBLE else View.GONE
            cardViewError.visibility = if (result) View.GONE else View.VISIBLE

        }

        override fun onProgressUpdate(vararg values: ProgressBeen?) {
            super.onProgressUpdate(*values)

            values[0]?.toCharSequence()?.let { list.add(it) }
            (mListView.adapter as ArrayAdapter<CharSequence>).notifyDataSetChanged()
        }

        override fun onCancelled() {
            super.onCancelled()
            progressDialog.dismiss()
        }
    }

}